### PR TITLE
Refactor Task class and its dependencies to match current version

### LINKED ISSUE #213: TodoList: Refactor Task class and its dependencies

### PR DESCRIPTION
**Pull request by michael-setia**
Fixes #213 

Added the Task class that will represent the task in the TodoList.

1. New file added :
  - Task class
  - Date class
  - Priority class (enum)
  - Type class (will be replaced with tags in the future)
  - TaskName class
  - Status class (enum)
------ [updated 17/10/2020] -------
2. The following files have also been added (taken from future PR) since I just realized they are also the base classes for TodoList:
  - DateComparator class
  - TaskComparatorByDate class
  - TaskComparatorByName class
  - TaskComparatorByPriority class
  - TaskNameComparator class
  - Criterion enum
  
3. The following files are updated:
  - TaskName class:
    - isSameTaskName() method. It now ignores the case for the string compared (uppercase or lowercase is the same)

  - Priority class:
    - Added message constraints
    - Added isValid() method() [not used at the moment]

### CODE REVIEW

File: src/main/java/seedu/address/model/todolist/TaskName.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        this.value = taskName;
+    }
+
+    /**
+     * Returns true if a given string is a valid task name.
+     *
+     * @param test given string
+     * @return true if task name is valid
+     */
+    public static boolean isValidTaskName(String test) {
```
Comment (jonasngs) [inline src/main/java/seedu/address/model/todolist/TaskName.java:30]: Perhaps this validation test could be modified to account for empty task name as well?

File: src/main/java/seedu/address/model/todolist/Type.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
+package seedu.address.model.todolist;
+
+/**
+ * Represents the type of the task.
+ */
+public enum Type {
```
Comment (jonasngs) [inline src/main/java/seedu/address/model/todolist/Type.java:6]: Perhaps you could extend a similar concept further and allow users to add tags to a particular task? The tagging feature seems to have some similarities with the Type enumeration here.

File: src/main/java/seedu/address/model/todolist/Task.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
+package seedu.address.model.todolist;
+
+/**
+ * Represents a Task in the todo list.
+ * Guarantees: details are present and not null, field values are validated, immutable.
+ */
```
Comment (jonasngs) [inline src/main/java/seedu/address/model/todolist/Task.java:6]: Will users be able to create Tasks without a type, priority or date? Perhaps allowing users to omit the addition of certain Task fields would increase flexibility?

**Review by jonasngs — APPROVED** [review]
The code is readable and understandable.  Apart from a few missing javadocs, it looks good to merge.

File: src/main/java/seedu/address/model/todolist/Task.java (hunk 3 above)
Comment (michael-setia) [reply src/main/java/seedu/address/model/todolist/Task.java]: Yes, I think we should do that. Will try to change in the next PR.

File: src/main/java/seedu/address/model/todolist/Type.java (hunk 2 above)
Comment (michael-setia) [reply src/main/java/seedu/address/model/todolist/Type.java]: Oh yes, I just realized that Tag can probably replace Type.

File: src/main/java/seedu/address/model/todolist/TaskName.java (hunk 1 above)
Comment (michael-setia) [reply src/main/java/seedu/address/model/todolist/TaskName.java]: Yup, totally forgot about that 😆

**Pull request merged**